### ISSUE TITLE
[DISCUSS] module metadata

### ISSUE THREAD

**Issue by LightGuard** [LightGuard/asciidoc-splitter#38]
We need to define some sort of metadata standard for modules so those modules can be properly split up during processing and contain the correct name and location. Unfortunately, we can not do this reliably using only the computer, a human will need to make some input.

My suggestion for this would be to add the information at the section level. Each entry should be prefixed with `module`, then a `_` for a separator, followed by the name of attribute. I don't have an exhaustive list of those, but my initial thoughts are `type` and perhaps a `used-in` attribute.

Here's an example of what my proposal would look like in asciidoc:
```
[module_type="con", module_used-in="getting-started,database"]
== Setting up a datasource
```

**Comment by maxandersen** [comment]
i kinda thought these would be attributes like:

== here is a section
:module-type: con
:module-used-in: "getting-started, database"

but if asciidoctor allow such arbitrary "annotations" on any kind of element ? then that works for me too.

I would use kebab case if that is allowed though (i.e. module-info rather than module_info)

btw. is "module_used" really relevant to put in ? is that actually used for something ?

**Comment by LightGuard** [comment]
**tl;dr**: Yes, those are supported and the preferred way to do this sort of thing.

I'm not married to the actual names as long as we standardize on something, kebab case is fine too.

`module-used`||`module_used` might be a nice to have. It does look like some of the documentation calls out in comments at the start of the module where it is used. Handy I suppose, but not sure if there's any other purpose behind that information.

There are document level attributes and element level attributes. What you're talking about above @maxandersen, are document level attributes, and are honestly a PITA to work with. Element level attributes live at the element level, naturally. Be that inline or block level, and are applied at the corresponding level, however, they are not inherited the way document level attributes seem to be (spoilers, they really aren't inherited either, they're more like global variables that can change depending on where you are in a document and you have to keep checking them to see what the current value is based on your location within the document, they also don't come through at the AST level unless we create a TreeProcessor extension and check at every level what the value is).

Attributes are essentially a dictionary. The [Asciidoctor User Manual section 10.5](https://asciidoctor.org/docs/user-manual/#setting-attributes-on-an-element) has good examples. One that we are all familiar with is the syntax for an image. Every bit in the square brackets is actually an attribute. In the AST they're both named and positional attributes for an image. A great example of attributes on a block is the listing or source block. The language, indent, and subs sections in the square brackets are all attributes. If you're itching to see what actually constitutes an attribute line, brush up your regex and take a look at https://github.com/asciidoctor/asciidoctor/blob/c94a24e2f4f5568375e45a0f7ae49c167a27b16d/lib/asciidoctor/rx.rb#L182. Should you feel especially curious, I recommend reading up on attribute parsing. You'll start at https://github.com/asciidoctor/asciidoctor/blob/c94a24e2f4f5568375e45a0f7ae49c167a27b16d/lib/asciidoctor/parser.rb#L2019 and eventually end up at https://github.com/asciidoctor/asciidoctor/blob/c94a24e2f4f5568375e45a0f7ae49c167a27b16d/lib/asciidoctor/attribute_list.rb#L24. The documentation for the AttributeList class has good examples and explains it quite well.

**Comment by sterobin** [comment]
@LightGuard , this looks interesting. Just wanted to comment that the naming convention for anchor IDs and file names are CCS-wide to comply with the Pantheon building tool chain. Today this is the template, as you've probably seen:

```
[id='proc_data-source-setting_{context}']
= Setting up a data source
```

where the file name would likewise be:

```
proc_data-source-setting.adoc
```

The hyphens and underscores in this template are significant, though currently being refined a little it seems. But today, this is the template.

The "module used in" comments are just an manually added aid so we know what all is impacted when we update the file, which can be important, especially with conditions, etc.

So bottom line, if it's a hard requirement to modify the naming and such for metadata purposes in the split, we just need to make sure that in the generated modules that result this template is adhere to. If manual maintenance is the only option, we'll have to see to what extent I suppose.

**Comment by LightGuard** [comment]
@sterobin if you’re talking about what Max was saying above with the hyphen vs underscore, he was simply indicating a preference for the attribute name. That would not have any influence on the generated name. 

As for manual maintenance, ideally the only thing is a human deciding what type of module it is. The rest we should be able to take care of during the split.

**Comment by LightGuard** [comment]
@sterobin as a follow-up, I doubt we’d have community contributors creating the IDs. Not sure if that matters.

**Comment by sterobin** [comment]
@LightGuard , regarding hyphens v underscores, right, understand what Max was saying. I was meaning something else, as in my example anchor IDs with `proc_module-name_{context}`. 

No, in reality, I don't expect contributors to follow the nuanced module templates, nor things like doc variables, etc. I think it's fair to have contributors try to be as organized as they can using other content as examples, but then fully expect for writers to maintain low-level requirements. That said, I've been very surprised at how well BA devs have jumped in cold and continued the trend using existing modules as templates.

**Comment by maxandersen** [comment]
it might be worth have the splitter be able to use id pattern as a fallback to identify module type.
Then it can be up to the individual communities which approach they will want: maintain id's or maintain module-type.

i.e. then the following three would be more or equivalent:
 
```
[id='proc_data-source-setting_{context}']
= Setting up a data source
```

```
[id='data-source-setting', module-type='proc']
= Setting up a data source
```

```
[module-type='proc']
= Setting up a data source
```

Thinking about it maybe just using id regexp like "([a-z][a-z][a-z])_(.*?)_{context}" as the pattern to look for might be better as then we don't need to swizzle id references in the generated output ?

**Comment by maxandersen** [comment]
> The "module used in" comments are just an manually added aid so we know what all is impacted when we update the file, which can be important, especially with conditions, etc.

Yeah, so i'm wondering why that needs to be there and why the Pantheon tool chain can't just keep track on what modules are used where so no need to have that manually and probably outofdate maintained. could we just drop it ?

**Comment by LightGuard** [comment]
Yes, we can drop it.

**Comment by LightGuard** [comment]
As of new we currently support splitting on

* The id containing `-{context}`
* Existence of the `module-type` attribute
* The id ending with `-proc`, `-con`, `-ref`
* The id beginning with `proc-`, `con-`, `ref-`

That probably covers everything? If so, (@sterobin, @maxandersen, @manaswinidas please comment) I move to close this issue.

**Comment by sterobin** [comment]
@LightGuard , sounds good to me.

**Comment by manaswinidas** [comment]
@LightGuard this covers all.
